src/CoaddTransmissionCurve.cc

```diff
@@ -0,0 +1,297 @@
+            _wavelengthBounds.second = std::max(_wavelengthBounds.second, inputWavelengthBounds.second);
+            // Set throughput at bounds from weighted average of throughput at bounds.
+            auto const inputThroughputAtBounds = transmission->getThroughputAtBounds();
+            _throughputAtBounds.first += inputThroughputAtBounds.first*weight;
+            _throughputAtBounds.second += inputThroughputAtBounds.second*weight;
+            // Add an element to the vector with all the stuff we need to store for each epoch.
+            CoaddInputData input = {
+                transmission,
+                record.getWcs(),
+                record.getValidPolygon(),
+                afw::geom::Box2D(record.getBBox()),
+                weight
+            };
+            _inputs.push_back(std::move(input));
+        }
+        _throughputAtBounds.first /= weightSum;
+        _throughputAtBounds.second /= weightSum;
+    }
+
+    // Private constructor used only for persistence
+    CoaddTransmissionCurve(
+        std::shared_ptr<afw::image::Wcs> coaddWcs,
+        std::pair<double, double> wavelengthBounds,
+        std::pair<double, double> throughputAtBounds,
+        std::vector<CoaddInputData> inputs
+    ) :
+        _coaddWcs(std::move(coaddWcs)),
+        _wavelengthBounds(std::move(wavelengthBounds)),
+        _throughputAtBounds(std::move(throughputAtBounds)),
+        _inputs(std::move(inputs))
+    {}
+
+    // All TransmissionCurves are immutable and noncopyable.
+    CoaddTransmissionCurve(CoaddTransmissionCurve const &) = delete;
+    CoaddTransmissionCurve(CoaddTransmissionCurve &&) = delete;
+    CoaddTransmissionCurve & operator=(CoaddTransmissionCurve const &) = delete;
+    CoaddTransmissionCurve & operator=(CoaddTransmissionCurve &&) = delete;
+
+    std::pair<double, double> getWavelengthBounds() const override {
+        return _wavelengthBounds;
+    }
+
+    std::pair<double, double> getThroughputAtBounds() const override {
+        return _throughputAtBounds;
+    }
+
+    void sampleAt(
+        afw::geom::Point2D const & position,
+        ndarray::Array<double const,1,1> const & wavelengths,
+        ndarray::Array<double,1,1> const & out
+    ) const override {
+        std::shared_ptr<afw::coord::Coord const> coord = _coaddWcs->pixelToSky(position);
+        ndarray::Array<double,1,1> tmp = ndarray::allocate(out.getShape());
+        tmp.deep() = 0.0;
+        out.deep() = 0.0;
+        double weightSum = 0.0;
+        for (auto const & input : _inputs) {
+            afw::geom::Point2D const inputPosition = input.sensorWcs->skyToPixel(*coord);
+            if (!input.bbox.contains(inputPosition)) {
+                continue;
+            }
+            if (input.validPolygon && !input.validPolygon->contains(inputPosition)) {
+                continue;
+            }
+            input.transmission->sampleAt(inputPosition, wavelengths, tmp);
+            tmp.deep() *= input.weight;
```

Comment:
Won't this always be zero? the array is defined as zero in all elements above, and you are using an in-place operator here...

Reply:
In the line above, `tmp` is filled by `sampleAt` (it's an output argument).